File: api/recurring.py
import services.recurring_service as service
from api.responses import ok, fail


def get_recurring_rules(user_id):
    try:
        return ok(service.get_recurring_rules(user_id))
    except ValueError as e:
        return fail(e)


def get_recurring_rule(recurring_rule_id):
    try:
        return ok(service.get_recurring_rule(recurring_rule_id))
    except ValueError as e:
        return fail(e)


def get_due_recurring_rules(as_of=None):
    try:
        return ok(service.get_due_recurring_rules(as_of=as_of))
    except ValueError as e:
        return fail(e)


def create_recurring_rule(account_id, category_id, name, expected_amount,
                          interval, frequency_unit, start_date, next_due_date,
                          end_date=None):
    try:
        return ok(
            service.create_recurring_rule(
                account_id=account_id,
                category_id=category_id,
                name=name,
                expected_amount=expected_amount,
                interval=interval,
                frequency_unit=frequency_unit,
                start_date=start_date,
                next_due_date=next_due_date,
                end_date=end_date,
            )
        )
    except ValueError as e:
        return fail(e)


def update_recurring_rule(recurring_rule_id, account_id=None, category_id=None,
                          name=None, expected_amount=None, interval=None,
                          frequency_unit=None, start_date=None, next_due_date=None,
                          end_date=None, status=None):
    try:
        kwargs = {k: v for k, v in {
            "account_id": account_id,
            "category_id": category_id,
            "name": name,
            "expected_amount": expected_amount,
            "interval": interval,
            "frequency_unit": frequency_unit,
            "start_date": start_date,
            "next_due_date": next_due_date,
            "end_date": end_date,
            "status": status,
        }.items() if v is not None}
        return ok(service.update_recurring_rule(recurring_rule_id, **kwargs))
    except ValueError as e:
        return fail(e)


def pause_recurring_rule(recurring_rule_id):
    try:
        return ok(service.pause_recurring_rule(recurring_rule_id))
    except ValueError as e:
        return fail(e)


def resume_recurring_rule(recurring_rule_id):
    try:
        return ok(service.resume_recurring_rule(recurring_rule_id))
    except ValueError as e:
        return fail(e)


def deactivate_recurring_rule(recurring_rule_id):
    try:
        return ok(service.deactivate_recurring_rule(recurring_rule_id))
    except ValueError as e:
        return fail(e)


def generate_transaction(recurring_rule_id):
    try:
        return ok(service.generate_transaction(recurring_rule_id))
    except ValueError as e:
        return fail(e)

File: frontend/src/lib/api/recurring.ts
import { callApi } from './client';

export type FrequencyUnit = 'day' | 'week' | 'month' | 'year';
export type RecurringStatus = 'active' | 'paused' | 'inactive';

export type RecurringRule = {
	recurring_rule_id: number;
	account_id: number;
	category_id: number;
	name: string;
	expected_amount: number | string;
	interval: number;
	frequency_unit: FrequencyUnit;
	start_date: string;
	next_due_date: string;
	end_date?: string | null;
	status: RecurringStatus;
};

export function getRecurringRules(userId: number) {
	return callApi<RecurringRule[]>('get_recurring_rules', userId);
}

export function getRecurringRule(recurringRuleId: number) {
	return callApi<RecurringRule | null>('get_recurring_rule', recurringRuleId);
}

export function getDueRecurringRules(asOf?: string | null) {
	return callApi<RecurringRule[]>('get_due_recurring_rules', asOf ?? null);
}

export function createRecurringRule(payload: {
	account_id: number;
	category_id: number;
	name: string;
	expected_amount: number | string;
	interval: number;
	frequency_unit: FrequencyUnit;
	start_date: string;
	next_due_date: string;
	end_date?: string | null;
}) {
	return callApi<{ recurring_rule_id: number }>(
		'create_recurring_rule',
		payload.account_id,
		payload.category_id,
		payload.name,
		payload.expected_amount,
		payload.interval,
		payload.frequency_unit,
		payload.start_date,
		payload.next_due_date,
		payload.end_date ?? null
	);
}

export function updateRecurringRule(
	recurringRuleId: number,
	payload: {
		account_id?: number | null;
		category_id?: number | null;
		name?: string | null;
		expected_amount?: number | string | null;
		interval?: number | null;
		frequency_unit?: FrequencyUnit | null;
		start_date?: string | null;
		next_due_date?: string | null;
		end_date?: string | null;
		status?: RecurringStatus | null;
	} = {}
) {
	return callApi<void>(
		'update_recurring_rule',
		recurringRuleId,
		payload.account_id ?? null,
		payload.category_id ?? null,
		payload.name ?? null,
		payload.expected_amount ?? null,
		payload.interval ?? null,
		payload.frequency_unit ?? null,
		payload.start_date ?? null,
		payload.next_due_date ?? null,
		payload.end_date ?? null,
		payload.status ?? null
	);
}

export function pauseRecurringRule(recurringRuleId: number) {
	return callApi<void>('pause_recurring_rule', recurringRuleId);
}

export function resumeRecurringRule(recurringRuleId: number) {
	return callApi<void>('resume_recurring_rule', recurringRuleId);
}

export function deactivateRecurringRule(recurringRuleId: number) {
	return callApi<void>('deactivate_recurring_rule', recurringRuleId);
}

export function generateTransaction(recurringRuleId: number) {
	return callApi<{ transaction_id: number }>('generate_transaction', recurringRuleId);
}
